cpp/src/arrow/util/async_generator.h

```diff
@@ -1063,6 +1063,86 @@ AsyncGenerator<T> MakeConcatenatedGenerator(AsyncGenerator<AsyncGenerator<T>> so
   return MergedGenerator<T>(std::move(source), 1);
 }
 
+template <typename T>
+struct Enumerated {
+  util::optional<T> value;
+  int index;
+  bool last;
+};
+
+template <typename T>
+struct IterationTraits<Enumerated<T>> {
+  static Enumerated<T> End() { return Enumerated<T>{{}, -1, false}; }
+  static bool IsEnd(const Enumerated<T>& val) { return !val.value.has_value(); }
+};
+
+/// \see MakeEnumeratedGenerator
+template <typename T>
+class EnumeratingGenerator {
+ public:
+  EnumeratingGenerator(AsyncGenerator<T> source, T initial_value)
+      : state_(std::make_shared<State>(std::move(source), std::move(initial_value))) {}
+
+  Future<Enumerated<T>> operator()() {
+    if (state_->finished) {
+      return AsyncGeneratorEnd<Enumerated<T>>();
+    } else {
+      auto state = state_;
+      return state->source().Then([state](const T& next) {
+        auto finished = IsIterationEnd<T>(next);
+        auto prev = Enumerated<T>{state->prev_value, state->prev_index, finished};
```

Comment:
Why are you buffering the previous value? Why not simply send the current one?

Reply:
I don't know if the current value is the last value or not.  All of this logic could be removed if I knew a priori how many batches were in a fragment.  However, for CSV I don't know this.  So if I'm sequencing batches from multiple CSV files and I see A1, A2, B1, B2, B3 does B1 come after A2?  Or is A3 just really slow in processing?

Reply:
Hmm, that doesn't seem to make sense. If you have several CSV files, you cannot reorder based on just the sequence number.

Reply:
When all is said and done what I have in the stream is something that looks like

fragment 2 batch 1 false // Buffer for resequencing
fragment 1 batch 1 false // Send downstream, first batch
fragment 1 batch 2 true // Send downstream, this is the next batch, then send f2b1 because I see this is the last batch in f1
fragment 2 batch 2 false // Send downstream, this follows the last batch we sent (f2b1)

Reply:
Perhaps, but you have a single index in your enumerating generator. How do you represent that?

Reply:
I start with `AsyncGenerator<Fragment>` which I enumerate to get `AsyncGenerator<Enumerated<Fragment>>`.

Next I map each `Enumerated<Fragment>`.  I call `Scan` to get `AsyncGenerator<RecordBatch>` I enumerate that to get `AsyncGenerator<Enumerated<RecordBatch>>`.  I then combine that with the `Enumerated<Fragment>` to get `AsyncGenerator<EnumeratedRecordBatch>` which is defined as...

```
struct EnumeratedRecordBatch {
  Enumerated<std::shared_ptr<RecordBatch>> record_batch;
  Enumerated<std::shared_ptr<Fragment>> fragment;
};
```

Each of these generators (one per fragment) are merged together which takes me from `AsyncGenerator<AsyncGenerator<EnumeratedRecordBatch>>` to `AsyncGenerator<EnumeratedRecordBatch>` which is exactly what I need (per ARROW-12288) to satisfy...

```
virtual Result<EnumeratedRecordBatchIterator> ScanBatchesUnordered();
```

I can then resequence that if the caller desires to get...

```
struct TaggedRecordBatch {
  std::shared_ptr<RecordBatch> record_batch;
  std::shared_ptr<Fragment> fragment;
};
```

By the method...

```
virtual Result<TaggedRecordBatchIterator> ScanBatches() = 0;
```

Reply:
I see. Well, if you need it, let's do it.